File: backend/src/main/java/com/leavemarker/entity/Plan.java
package com.leavemarker.entity;

import com.leavemarker.enums.BillingCycle;
import com.leavemarker.enums.PlanTier;
import com.leavemarker.enums.PlanType;
import jakarta.persistence.*;
import lombok.*;

import java.math.BigDecimal;

@Entity
@Table(name = "plans")
@Getter
@Setter
@NoArgsConstructor
@AllArgsConstructor
@Builder
public class Plan extends BaseEntity {

    @Column(nullable = false, length = 100)
    private String name;

    @Column(length = 500)
    private String description;

    @Enumerated(EnumType.STRING)
    @Column(nullable = false)
    private PlanTier tier;

    @Enumerated(EnumType.STRING)
    @Column(nullable = false)
    @Builder.Default
    private PlanType planType = PlanType.FREE;

    @Enumerated(EnumType.STRING)
    @Column(nullable = false)
    private BillingCycle billingCycle;

    // Pricing
    @Column(nullable = false, precision = 10, scale = 2)
    @Builder.Default
    private BigDecimal monthlyPrice = BigDecimal.ZERO;

    @Column(nullable = false, precision = 10, scale = 2)
    @Builder.Default
    private BigDecimal yearlyPrice = BigDecimal.ZERO;

    // Employee limits
    @Column(nullable = false)
    @Builder.Default
    private Integer minEmployees = 1;

    @Column(nullable = false)
    @Builder.Default
    private Integer maxEmployees = 10;

    // Leave policy limits
    @Column(nullable = false)
    @Builder.Default
    private Integer maxLeavePolicies = 1; // FREE: 1, MID_TIER: unlimited (-1)

    // Holiday limits
    @Column(nullable = false)
    @Builder.Default
    private Integer maxHolidays = 6; // FREE: 6, MID_TIER: unlimited (-1)

    @Column(nullable = false)
    @Builder.Default
    private Boolean active = true;

    // Feature flags
    @Column(nullable = false)
    @Builder.Default
    private Boolean attendanceManagement = false; // MID_TIER only

    @Column(nullable = false)
    @Builder.Default
    private Boolean reportsDownload = false; // MID_TIER only

    @Column(nullable = false)
    @Builder.Default
    private Boolean multipleLeavePolicies = false; // MID_TIER only

    @Column(nullable = false)
    @Builder.Default
    private Boolean unlimitedHolidays = false; // MID_TIER only

    @Column(nullable = false)
    @Builder.Default
    private Boolean attendanceRateAnalytics = false; // MID_TIER only

    @Column(nullable = false)
    @Builder.Default
    private Boolean advancedReports = false; // MID_TIER only

    @Column(nullable = false)
    @Builder.Default
    private Boolean customLeaveTypes = false; // MID_TIER only

    @Column(nullable = false)
    @Builder.Default
    private Boolean apiAccess = false; // MID_TIER only

    @Column(nullable = false)
    @Builder.Default
    private Boolean prioritySupport = false; // MID_TIER only

    // Pricing per employee
    @Column(nullable = false, precision = 10, scale = 2)
    @Builder.Default
    private BigDecimal pricePerEmployee = BigDecimal.ZERO;

    // Report download pricing (add-on for MID_TIER)
    @Column(precision = 10, scale = 2)
    @Builder.Default
    private BigDecimal reportDownloadPriceUnder50 = BigDecimal.ZERO; // ₹200/month for <50 employees

    @Column(precision = 10, scale = 2)
    @Builder.Default
    private BigDecimal reportDownloadPrice50Plus = BigDecimal.ZERO; // ₹400/month for >=50 employees
}

File: backend-go/internal/models/plan.go
package models

import (
	"github.com/leavemarker/backend-go/internal/enums"
	"github.com/shopspring/decimal"
)

// Plan represents a subscription plan entity
type Plan struct {
	BaseModel
	Name                        string              `gorm:"size:100;not null" json:"name"`
	Description                 string              `gorm:"size:500" json:"description"`
	Tier                        enums.PlanTier      `gorm:"size:20;not null" json:"tier"`
	PlanType                    enums.PlanType      `gorm:"size:20;default:FREE" json:"planType"`
	BillingCycle                enums.BillingCycle  `gorm:"size:20" json:"billingCycle"`
	MonthlyPrice                decimal.Decimal     `gorm:"type:decimal(10,2);default:0.00" json:"monthlyPrice"`
	YearlyPrice                 decimal.Decimal     `gorm:"type:decimal(10,2);default:0.00" json:"yearlyPrice"`
	MinEmployees                int                 `gorm:"default:1" json:"minEmployees"`
	MaxEmployees                int                 `gorm:"default:10" json:"maxEmployees"`
	MaxLeavePolicies            int                 `gorm:"default:1" json:"maxLeavePolicies"`
	MaxHolidays                 int                 `gorm:"default:6" json:"maxHolidays"`
	Active                      bool                `gorm:"default:true" json:"active"`

	// Feature flags
	AttendanceManagement        bool                `gorm:"default:false" json:"attendanceManagement"`
	ReportsDownload             bool                `gorm:"default:false" json:"reportsDownload"`
	MultipleLeavePolicies       bool                `gorm:"default:false" json:"multipleLeavePolicies"`
	UnlimitedHolidays           bool                `gorm:"default:false" json:"unlimitedHolidays"`
	AttendanceRateAnalytics     bool                `gorm:"default:false" json:"attendanceRateAnalytics"`
	AdvancedReports             bool                `gorm:"default:false" json:"advancedReports"`
	CustomLeaveTypes            bool                `gorm:"default:false" json:"customLeaveTypes"`
	APIAccess                   bool                `gorm:"default:false" json:"apiAccess"`
	PrioritySupport             bool                `gorm:"default:false" json:"prioritySupport"`

	// Pricing
	PricePerEmployee            decimal.Decimal     `gorm:"type:decimal(10,2);default:0.00" json:"pricePerEmployee"`
	ReportDownloadPriceUnder50  decimal.Decimal     `gorm:"type:decimal(10,2);default:0.00" json:"reportDownloadPriceUnder50"`
	ReportDownloadPrice50Plus   decimal.Decimal     `gorm:"type:decimal(10,2);default:0.00" json:"reportDownloadPrice50Plus"`

	// Relationships
	Subscriptions []Subscription `gorm:"foreignKey:PlanID" json:"-"`
	Payments      []Payment      `gorm:"foreignKey:PlanID" json:"-"`
}

func (Plan) TableName() string {
	return "plans"
}
